ESPS/general/src/vq/vq.1
.\" @(#)vq.1	3.9 4/7/97 ESI
.\" Copyright (c) 1987 Entropic Speech, Inc.; All rights reserved
.TH VQ 1\-ESPS 4/7/97
.ds ]W "\fI\s+4\ze\h'0.05'e\s-4\v'-0.4m'\fP\(*p\v'0.4m'\ Entropic Speech, Inc.
.SH "NAME"
vq - vector quantize a field in FEA records
.SH "SYNOPSIS"
.B vq
[
.BI \-x " debug_level"
] [
.BI \-h " histfile"
] [
.BI \-f " fieldname"
] [
.BI \-c " record"
] [
.BI \-i 
]
.I "cbk.fea fea.in fea.out"
.SH "DESCRIPTION"
.PP
.I Vq
reads the ESPS FEA_VQ file 
.I cbk.fea
to obtain a vector quantization codebook.  
.I Vq
then reads an file
.I fea.in,
encodes a particular field from each FEA record, and then 
outputs the encoded fields to
.I fea.out.
.I Fea.out
also contains the rest of the information from 
.I fea.in,
i.e., the contents of the two files are identical except for the 
quantized field within each record.  
.PP
If "\-" is supplied in place of 
.I fea.in, 
then standard input is used.  
If "\-" is supplied in place of 
.I fea.out,
then standard output is used.  
.PP
.I Vq
determines which distortion measure to use from the information 
stored in the FEA_VQ file.  In many cases, 
.I vq
will be used with codebooks designed using 
.I vqdes
(1\-ESPS), although this is not required.  If there are field 
to be quantized in 
.I input.ana
is larger than the dimension of the codewords in 
.I cbk.fea,
the extra elements are set to zero in the output file.  
.PP
The field to quantize (specified by the \fB-f\fP option) must be of
type FLOAT.  Note that \fIfeafunc\fP can be used to transform the data
type of fields.  
.SH OPTIONS
.PP
The following options are supported:
.TP
.BI \-x " debug_level"
If 
.I "debug_level \fR[0]\fP"
is positive,
.I vq
outputs debugging messages and other information about the codebook
design to a history file.  The messages proliferate as 
.I debug_level 
increases.  For level 0, there is no output.  For level 1, the size of
the input sequence, the average distortion, the standard deviation of
the distortion, and the maximum distortion are output.  For level 2, the
codebook itself is also output.  For level 3, information is
output for every input vector, namely the index of the closest codeword,
the distortion with respect to the closest codeword,
and the cumulative average distortion.  For level 4, the output also
includes the input vectors and the output vectors.
.TP
.BI \-h " histfile \fR[vqhist]\fP"
Specifies the file to use for history and debugging output.  
.TP
.BI \-f " fieldname \fR[spec_param]\fP"
Specifies the name of the field to be quantized in each 
.I fea.in record.  
.TP
.BI \-c " record"
Specifies an integer record number within 
.I cbk.fea
to use as the codebook.  If the \fB\-c\fP option is not used, 
then the last record in 
.I cbk.fea
is used as the codebook.  
.TP 
.BI \-i
Specifies that only the codword index and not the quantized
field \fIfieldname\fP be written to \fIfea.out\fP. 
.SH "ESPS PARAMETERS"
.PP
The ESPS parameter file is not read.  
.SH ESPS COMMON
.PP
No values are written to ESPS Common.  
.SH ESPS HEADERS
.PP
.I Vq
reads the usual information from the \fIfea.in\fP header.  The header
for \fIfea.out\fP is created using 
.I copy_header(3\-ESPS),
with the usual record-keeping fields filled in.  A field is added
which contains the codeword index associated with the quantized value
of \fIfieldname\fP. The name of this field is obtained by appending
the suffix \fI_cwndx\fP to the name of the quantized field, e.g.
the default field name is \fIspec_param_cwndx\fP.  This field is a 
single integer of data type LONG.  If the \fB\-i\fP option is
specified, the field \fIfieldname\fP is not included in the file
\fIfea.out\fP.
.PP
The generic header item \fIstart_time\fP is written in the output
file.  The value written is computed by taking the \fIstart_time\fP
value from the header of the input file (or zero, if such a header
item doesn't exist) and adding to it the relative time from the first
record in the file to the first record processed.  This relative time
is computed using the generic \fIrecord_freq\fP from the input file.
If \fIrecord_freq\fP doesn't exist in the input file, \fIstart_time\fP
is copied from the input file to the output file.
.PP
If it exists in the input file header, the generic header item
\fIrecord_freq\fP is copied to the output file header.  This item
gives the number of records per second of original data analyzed.
.PP
The generic header item \fIquantized\fP is set to YES in the output file.  
.PP
Both 
.I cbk.fea
and
.I fea.in
are added to the output header as sources.  
.PP
.I Vq
creates and writes the following generic header items (names
incremented using 
.I uniq_name
(3\-ESPS) if they exist already):
.sp
.I "quantized_field"
.PP
This is the name of the quantized field (from \fB\-f\fP option).
.sp
.I "cbk_current_size"
.PP
The size of the quantization codebook.
.sp
.I "cbk_dimen"
.PP
The dimension of codewords in the quantization codebook.
.sp
.I "cbk_dist_type"
.PP
The type of distortion used in designing the codebook and in
quantizing 
.I fea.in.
.sp
.I "cbk_cbk_struct"
.PP
The structure of the quantization codebook.
.sp
.I "cbk_cbk_type"
.PP
The type of the quantization codebook.
.sp
.I "cbk_final_dist"
.PP
The final distortion of the training sequence that was used in the
original design of the quantization codebook.
.sp
.I "encode_distortion"
.PP
The average distortion resulting from encoding 
.I fea.in.
This is left as -1 if the output is standard output.  
.SH "FUTURE CHANGES"
.PP

.SH "SEE ALSO"
.PP
.nf
vqdes(1\-ESPS), vqdesasc(1\-ESPS), vqasc(1\-ESPS), 
vqdesign(3\-ESPS), \fIfeafunc\fP(1\-\s-1ESPS\s+1), \fImergefea\fP (1\-\s-1ESPS\s+1),
FEA_VQ(5\-ESPS), FEA(5\-ESPS), ESPS(5\-ESPS)
.fi
.SH "WARNINGS AND DIAGNOSTICS"
.PP
.PP
Vq will exit with a diagnostic if any of the following hold:
.IP
It can't read the input ESPS file headers or they are not of the right type.
.IP
.I Fieldname 
is not a valid record field in \fIfea.in\fP.
.IP
The field 
.I fieldname
in \fIfea.in\fP is not of type FLOAT.
.IP
The field 
.I fieldname
in \fIfea.in\fP is smaller than the size of the VQ codewords.
.IP
There are insufficient records in \fIcbk.fea\fP.  
.PP
Warnings will be issued if any of the following hold:
.IP
The field 
.I fieldname
in \fIfea.in\fP is larger than the size of the VQ codewords. 
.IP
The field
.I fieldname
in \fIfea.in\fP has a different name than the contants generic header item
\fIfea_field\fP in \fIcbk.fea\fP, if that generic header item exists. 
.IP
The generic header item \fIquantized\fP is YES in \fIfea.in\fP (i.e., 
if any quantization has occured already on the input file).  
.SH "BUGS"
.PP
None known.  
.SH "REFERENCES"
.PP
See
.I vqdesign
(3\-ESPS).  
.SH "AUTHOR"
.PP
Manual page by John Shore; program by John Shore.  Codeword index
field added by Bill Byrne.




ESPS/general/src/vq/vq.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1986-1990  Entropic Speech, Inc. 
 *    "Copyright (c) 1990-1996  Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  John Shore
 * Checked by:
 * Modification History: 
 *   1/8/91: -i option and codword index field added by Bill Byrne.
 *
 * Brief description:  This program is used to vector-quantize FEA fields.
 */

static char *sccs_id = "@(#)vq.c	3.14	1/20/97	ESI/ERL";

/*
 * include files
 */
#include <stdio.h>
#include <string.h>
#include <esps/esps.h>
#include <esps/fea.h>
#include <esps/vq.h>
/*
 * defines
 */
#define MAXLINE 200
#define MAXCMNT 2047  /* Number of characters in user entered comments. */
#define Fprintf (void)fprintf
#define Fflush (void)fflush
#define ERROR_EXIT(text) {(void) fprintf(stderr, "%s: %s - exiting\n", \
		ProgName, text); exit(1);}
#define SYNTAX USAGE ("vq [-x debug_level] [-h histfile] [-c record] [-f fieldname] [-i] cbk.fea fea.in fea.out")
#define FEA_BUFSIZE 500
/*
 * system functions and variables
 */
#ifndef IBM_RS6000
int getopt();
double	atof();
long time();
int atoi(), fclose();
char *strtok();
char *strcat();
char *fgets();
void rewind();
char *mktemp(), *ctime();
void rewind();
#if !defined(DEC_ALPHA) && !defined(LINUX_OR_MAC)
char *malloc();
char *calloc();
#endif
double sqrt();
#endif
extern  optind;
extern	char *optarg;
/*
 * external SPS functions
 */
char *uniq_name();
char *get_cmd_line();
long *add_genhd_l();
short *add_genhd_s();
float *add_genhd_f();
char *add_genhd_c();
char *get_genhd();
int genhd_type();
int getsym_i(), lin_search();
struct header *read_header();
int rc_to_lar();
long vqencode();
char *savestring();
int add_fea_fld();
int del_fea_fld();
/*
 * global function declarations
 */
void vec_cy_f();
double mse_lar_dist();
void wrt_fea_vec();
/*
 * global variable declarations
 */
char		    *ProgName = "vq";
char		    *histfile = "vqhist";	/*output ASCII file for history data*/
FILE		    *histrm;
char		    *local_no_yes[] = {"NO", "YES", NULL}; 

int		debug_level = 0; /* level of history detail;
				  global for library access*/

/*
 * main program
 */
main (argc, argv)
int argc;
char **argv;
{
/*
 * setup and initialization
 */
    char		*Version = "3.14";
    char		*Date = "1/20/97";
    char		*cbk_fea = NULL;    /*input FEA_VQ file for codebook*/
    struct header	*cbk_ih;
    FILE		*cbk_strm;	    
    char		*data_in;	    /*input FEA file for quantization*/
    FILE		*in_strm = stdin;   
    struct header	*fea_ih;
    struct fea_data	*fea_rec;
    struct fea_data	*fea_rec_out;
    char		*data_out;	    /*output FEA file - quantized*/
    struct header	*fea_oh;
    FILE		*out_strm = stdout;
    char		*cmd_line;	    /*to hold command line*/
    int			c;		    /*for getopt return*/
    long		tloc;		    /*for local time*/
    long		fea_dim;	    /*numb. columns in feadata*/
    struct vqcbk	*cbk;		    /*the VQ codebook*/
    double		(*distort)();	    /*routine to use for distortion*/
    float		*input_vec;	    /*one feature vector*/
    long		nfea;		    /*number of feature vectors*/
    long		curr_ind;	    /*current feature vector*/
    long		max_ind;	    /*vector index of max distortion*/
    double		curr_dist;	    /*distortion of current vector*/
    double		max_dist;	    /*maximum vector distortion*/
    double		tot_sqdist;	    /*cumulative square distortion*/
    double		tot_dist;	    /*cumulative distortion*/
    double		dist_mean;	    /*mean distortion*/
    double		dist_variance;	    /*variance of distortion*/
    double		dist_sd;	    /*standard deviation of distortion*/
    int			dist_err = 0;	    /*error return from distort*/
    int			ncbkrec;	    /*record number of initial codebook*/
    int			c_flag = 0;	    /*flag for -c option*/
    long		i;
    float		*field;		    /*pointer to field to be quantized*/
    float		*field_out;	    /*pointer to quantized field in output*/
    char		*fieldname = "spec_param"; /*name of FEA field*/
    char		*enc_dist_name = "encode_distortion"; /*a generic name*/
    char		*tempname;	    /*holder for generic name*/
    long		fea_size;	    /*size of FEA field data*/
    int			dumnumb;	    /*dummy variable for number of generics */
    char		*cbk_field;	    /*name of design field*/
    int	    		sizedummy;	    /*dummy variable for genhd_type*/
    int                 i_flag = 0;         /*flag for -i option*/
    char                *ndx_fieldname;     /*field name for code word index*/
    long                *cw_ndx=NULL;       /*pointer to code word index*/
/*
 *Initialization
 */
    cmd_line = get_cmd_line(argc, argv);    /*store copy of command line*/    
/*
 * process command line options
 */

    while ((c = getopt (argc, argv, "x:h:c:f:i")) != EOF) {
	switch (c) {
	    case 'x': 
		debug_level = atoi (optarg);
		break;
	    case 'h':
		histfile = optarg;
		break;
	    case 'c':
		ncbkrec = atoi(optarg);
		c_flag++;
		break;
	    case 'f':
		fieldname = optarg;
 		break;
            case 'i':
                i_flag++;
		break;
	    default:
		SYNTAX;
	}
    }
    
/*
 * process file arguments 
 */
    if (optind < argc) {
	cbk_fea = argv[optind++];
	if (strcmp (cbk_fea, "-") == 0)
	    ERROR_EXIT("vq: can't use - for codebook -- exiting");
        TRYOPEN (argv[0], cbk_fea, "r", cbk_strm);
    }
    else {
	Fprintf(stderr, "vq: no FEA codebook file specified.\n");
	SYNTAX;
        
    }
    if (optind < argc) {
	data_in = argv[optind++];
	if (strcmp (data_in, "-") == 0)
	    data_in = "<stdin>";
	else
	    TRYOPEN (argv[0], data_in, "r", in_strm);
	
    }
    else {
	Fprintf(stderr, "vq: no input data file specified.\n");
	SYNTAX;
    }

    if (optind < argc) {
	data_out = argv[optind++];
	if (strcmp (data_out, "-") == 0)
	    data_out = "<stdout>";
	else
	    TRYOPEN (argv[0], data_out, "w", out_strm);
    }
    else {
	Fprintf(stderr, "vq: no output data file specified.\n");
	SYNTAX;
    }

    TRYOPEN(argv[0], histfile, "w", histrm);
/*
 * read FEA file and get codebook
 */

    if ((cbk_ih = read_header(cbk_strm)) == NULL)
	NOTSPS(argv[0], cbk_fea);
    if (cbk_ih->common.type != FT_FEA) 
	ERROR_EXIT("codebook file not FEA file");
    if (cbk_ih->hd.fea->fea_type != FEA_VQ) 
	ERROR_EXIT("codebook file not FEA_VQ file");
    cbk = allo_vqfea_rec(cbk_ih);
    assert (cbk != NULL);

    if (!c_flag)
	ncbkrec = cbk_ih->common.ndrec;

    if (ncbkrec >= 1)
    {
	fea_skiprec(cbk_strm, (long) ncbkrec - 1, cbk_ih);

	if (get_vqfea_rec(cbk, cbk_ih, cbk_strm) == EOF)
	    ERROR_EXIT("EOF when trying to read initial codebook");
    }
    else if (ncbkrec == -1)
    {
	/* Variable-length records (e.g. Esignal Ascii format).
	 * Read records to find the last.
	 */
	struct vqcbk     *next_cbk, *tmp_cbk;

	next_cbk = allo_vqfea_rec(cbk_ih);

	if (get_vqfea_rec(cbk, cbk_ih, cbk_strm) == EOF)
	    ERROR_EXIT("EOF when trying to read initial codebook");

	while (get_vqfea_rec(next_cbk, cbk_ih, cbk_strm) != EOF)
	{
	    tmp_cbk = cbk;
	    cbk = next_cbk;
	    next_cbk = tmp_cbk;
	}
    }
    else
	ERROR_EXIT("codebook record number less than 1");

    fea_dim = *cbk->dimen;

/*
 *select distortion function
 */
    switch (*cbk->dist_type) {
	case MSE:  
	    distort = NULL;
	    break;
	case MSE_LAR:
	    distort = mse_lar_dist;
	    break;
	default:
	    ERROR_EXIT("invalid distortion type");
    }
/*
 * initial history and debug output 
 */

    if (debug_level > 0) {
        tloc = time(0); 
        Fprintf(histrm, "vq history ouput on %s", ctime(&tloc));
        Fprintf(histrm, "vq version %s, date %s\n", Version, Date); 
        Fprintf(histrm, "command line:\n%s", cmd_line);
	Fprintf(histrm, "codebook type: %s\n", vq_cbk_types[*cbk->cbk_type]);
	Fprintf(histrm, "codebook structure: %s\n", cbk_structs[*cbk->cbk_struct]);
	Fprintf(histrm, "codebook distortion: %s\n", dist_types[*cbk->dist_type]);
	Fprintf(histrm, "codebook dimension: %d\n", *cbk->dimen);
	Fprintf(histrm, "codebook size: %d\n", *cbk->current_size);	
    }	
    if (debug_level > 1) {
	Fprintf(histrm, "\nCodebook:\n\n");
	print_fea_rec(cbk->fea_rec, cbk_ih, histrm);
    }

/*
 * read FEA input header, allocate FEA record and temp space
 */
    if ((fea_ih = read_header(in_strm)) == NULL)      
	ERROR_EXIT("couldn't read input FEA file header");
    if (fea_ih->common.type != FT_FEA)
	ERROR_EXIT("file to encode not FEA file");
    /* if input not a pipe, can check to make sure data is there*/
    if (fea_ih->common.ndrec == 0) 
	ERROR_EXIT("no data in input file\n");
    fea_rec = allo_fea_rec(fea_ih);
    input_vec = (float *) calloc(fea_dim, sizeof (float));
/*
 * consistency checks
 */
    fea_size = get_fea_siz(fieldname, fea_ih, (short *) NULL, (long **) NULL);
    if (fea_size == 0) {
	Fprintf(histrm, 
	    "vq: field %s doesn't exist in input FEA file\n", fieldname);
	Fprintf(stderr, 
	    "vq: field %s doesn't exist in input FEA file\n", fieldname);
	exit(1);
    }
    if (fea_dim > fea_size) ERROR_EXIT("codeword size larger than FEA field size");
    if (fea_dim != fea_size)  {
	Fprintf(histrm, 
	  "vq: WARNING - size of FEA field greater than codewords\n");
	Fprintf(stderr, 
	  "vq: WARNING - size of FEA field greater than codewords\n");
    }
    if (get_fea_type(fieldname, fea_ih) != FLOAT) 
	ERROR_EXIT("specified field is not type FLOAT");
    if (get_genhd("quantized", fea_ih) != NULL &&
	    (*(short *) get_genhd("quantized", fea_ih)) == YES) {
	Fprintf(histrm, 
	    "vq: WARNING - input file already has some quantized fields\n");
	Fprintf(stderr, 
	    "vq: WARNING - input file already has some quantized fields\n");
    }
    /*
     * check for consistency between quantization field and codebook design
     */
    if (genhd_type("fea_field", &sizedummy, cbk_ih) != HD_UNDEF) {
	if (strcmp(get_genhd("fea_field", cbk_ih), fieldname) != 0) {
	    cbk_field = get_genhd("fea_field", cbk_ih);
	    Fprintf(histrm, 
		"vq: WARNING - inconsistent fea_field in codebook\n");
	    Fprintf(histrm, "\tinput field being quantized: %s\n", fieldname);
	    Fprintf(histrm, 
		"\tcodebook designed from field named: %s\n", cbk_field);
	    Fprintf(stderr, 
		"vq: WARNING - inconsistent fea_field in codebook\n");
	    Fprintf(stderr, "\tinput field being quantized: %s\n", fieldname);
	    Fprintf(stderr, 
		"\tcodebook designed from field named:  %s\n", cbk_field);
	}
    }
    else {
	Fprintf(histrm, "vq: WARNING - codebook has no record of design fieldname\n");
	Fprintf(stderr, "vq: WARNING - codebook has no record of design fieldname\n");	
    }
    /*
     *create and define header of output FEA file
     */
    fea_oh = copy_header(fea_ih);
    (void) strcpy (fea_oh->common.prog, "vq");
    (void) strcpy (fea_oh->common.vers, Version);
    (void) strcpy (fea_oh->common.progdate, Date);
    add_source_file(fea_oh, cbk_fea, cbk_ih);
    add_source_file(fea_oh, data_in, fea_ih);
    add_comment(fea_oh, cmd_line);
    /*
     * fill "quantized" field if there (FEA_ANA), else create it
     */
    if (get_genhd("quantized", fea_ih) != NULL)
	*(short *) get_genhd("quantized", fea_oh) = YES;
    else
    *add_genhd_e("quantized", (short *) NULL, 1, local_no_yes, fea_oh) = YES;
    /*
     * add generic header items that describe codebook used and quantized
     * field
     */
    tempname = uniq_name("quantized_field", genhd_list(&dumnumb, fea_oh));
    (void) add_genhd_c(tempname, fieldname, strlen(fieldname), fea_oh);
    tempname = uniq_name("cbk_current_size", genhd_list(&dumnumb, fea_oh));
    (void) add_genhd_l(tempname, cbk->current_size, 1, fea_oh);
    tempname = uniq_name("cbk_dimen", genhd_list(&dumnumb, fea_oh));
    (void) add_genhd_l(tempname, cbk->dimen, 1, fea_oh);
    tempname = uniq_name("cbk_dist_type", genhd_list(&dumnumb, fea_oh));
    (void) add_genhd_e(tempname, cbk->dist_type, 1, dist_types, fea_oh);
    tempname = uniq_name("cbk_cbk_struct", genhd_list(&dumnumb, fea_oh));
    (void) add_genhd_e(tempname, cbk->cbk_struct, 1, cbk_structs, fea_oh);
    tempname = uniq_name("cbk_cbk_type", genhd_list(&dumnumb, fea_oh));
    (void) add_genhd_e(tempname, cbk->cbk_type, 1, vq_cbk_types, fea_oh);
    tempname = uniq_name("cbk_final_dist", genhd_list(&dumnumb, fea_oh));
    (void) add_genhd_d(tempname, cbk->final_dist, 1, fea_oh);
    enc_dist_name = uniq_name("encode_distortion", genhd_list(&dumnumb, fea_oh));
    *add_genhd_d(enc_dist_name, (double *) NULL, 1, fea_oh) = -1;

    if ( (ndx_fieldname = (char *) malloc( (unsigned)(strlen(fieldname) + 10) )) == NULL) {
	Fprintf(stderr, "%s: can't allocate field name memory.\n", ProgName);
	exit(1);
    }
    strcpy(ndx_fieldname, fieldname);
    strcat(ndx_fieldname, "_cwndx");

    if (-1 == add_fea_fld(ndx_fieldname, 1L, 0,
			  (long *) NULL, LONG, (char **) NULL, fea_oh)) {
	Fprintf(stderr, "trouble adding %s to output record.\n", ndx_fieldname);
	exit(1);
    }
    if (i_flag) {
	if (del_fea_fld(fieldname, fea_oh)==-1) {
	    Fprintf(stderr, "%s: can't remove %s.\n", ProgName, fieldname);
	    exit(1);
	}
    }

    update_waves_gen(fea_ih, fea_oh, 1.0, 1.0);
    write_header(fea_oh, out_strm);
    fea_rec_out = allo_fea_rec(fea_oh);

    if ( (field = (float *) get_fea_ptr(fea_rec, fieldname, fea_ih)) == NULL) {
	Fprintf(stderr, "can't get field %s in input file.\n", fieldname);
	exit(1);
    }
    if ( (cw_ndx = (long *) get_fea_ptr(fea_rec_out, ndx_fieldname, fea_oh)) == NULL) {
	Fprintf(stderr, "can't get field %s.\n", ndx_fieldname);
	exit(1);
    }
    if (!i_flag) 
	if ( (field_out = (float *) get_fea_ptr(fea_rec_out, fieldname, fea_oh)) == NULL) {
	    Fprintf(stderr, "can't get field %s in output file.\n", fieldname);
	    exit(1);
	}

/*
 * main processing goes here
 */
    nfea = max_dist = tot_dist = tot_sqdist = 0.0;
    while (get_fea_rec(fea_rec, fea_ih, in_strm) != EOF) {
	nfea++;
	vec_cy_f(input_vec, field, fea_dim);
	/*
	 *encode the vector
	 */
        switch (*cbk->cbk_struct) {
	    case FULL_SEARCH:  
		curr_ind = vqencode(input_vec, fea_dim, cbk->codebook, 
		  *cbk->current_size, *cbk->dimen, &curr_dist, distort, &dist_err);
		break;
	    default:
		ERROR_EXIT("invalid cbk_struct type");
	}
	/* copy codeword index to output field */
	*cw_ndx = curr_ind;
	/*Write out the encoded vector, making sure to zero 
	 *the refl. coefficients beyond the codebook dimension
         */
	copy_fea_rec( fea_rec, fea_ih, fea_rec_out, fea_oh, (char **)NULL, (short **)NULL);
	if (!i_flag) {
	    vec_cy_f(field_out, cbk->codebook[curr_ind], fea_dim);
	    for (i = fea_dim; i < fea_size; i++) 
		field_out[i] = 0.0;
	}
	put_fea_rec(fea_rec_out, fea_oh, out_strm);
	/*
	 *history stuff
 	 */
	tot_dist += curr_dist;
	tot_sqdist += curr_dist * curr_dist;
	if (debug_level > 2) {
	    Fprintf(histrm, "vq: encoding input vector %d\n", nfea);
	    if (debug_level > 3) {
		Fprintf(histrm, "     Input vector:\n");
		wrt_fea_vec(histrm, input_vec, *cbk->dimen);
	    }
	    Fprintf(histrm, 
		"     Encoding index = %d, distortion = %g\n",  curr_ind, curr_dist);
	    Fprintf(histrm, 
		"     Current average distortion = %g\n", tot_dist / nfea);
	    if (debug_level > 3) {
		Fprintf(histrm, "     Encoded vector:\n");
		wrt_fea_vec(histrm, cbk->codebook[curr_ind], *cbk->dimen);
	    }
	    Fflush(histrm);
	}
	if (curr_dist > max_dist) {
	    max_dist = curr_dist;
	    max_ind = curr_ind;
	}
	if (dist_err) ERROR_EXIT("error return from distort function");
    }
    /*  compute mean encoding distortion and write it back in the header
     */
    if (out_strm != stdout) {
	dist_mean = tot_dist / nfea;	
	*(double *)get_genhd(enc_dist_name, fea_oh) = dist_mean;
	rewind(out_strm);
	write_header(fea_oh, out_strm);
    }
    /*
     * write some summary statistics to the history file
     */
    if (debug_level > 0) {
	dist_variance = (tot_sqdist / nfea) - dist_mean * dist_mean;
	dist_sd = sqrt(dist_variance);
	Fprintf(histrm, "vq: encoded %d input vectors\n", nfea);
	Fprintf(histrm, "\tmean distortion = %g, stand. dev. = %g\n", dist_mean, dist_sd);
	Fprintf(histrm, "\tmaximum distortion was %g at vector %d\n", max_dist, max_ind);
	Fflush(histrm);
    }
    exit(0);
}

double
mse_lar_dist(fea_vector, vec_dim, codeword, cdwd_dim, error)
float *fea_vector;
long vec_dim;
float *codeword;
long cdwd_dim;
int *error;
/*
 *This distortion function assumes both the feature vector and 
 *the codeword are reflection coefficients; it converts both to 
 *log area ratios and then does mean-square error.  
 */
{
    double diff;
    float fea_lar, cd_lar;
    double msedist = 0.0;
    while (vec_dim-- != 0) {
    	(void) rc_to_lar(*fea_vector++, &fea_lar);
	(void) rc_to_lar(*codeword++, &cd_lar);
	diff = fea_lar - cd_lar;
	msedist += diff * diff;
    }
    return msedist / cdwd_dim;
}    


void
wrt_fea_vec(strm, fea_vec, size)
FILE *strm;		/*file stream to use*/
float *fea_vec;		/*feature vector to print*/
long size;	/*dimension of feature vector*/
/*
 *utility function for printing a feature vector*/
{
    long i;
    for (i = 0; i < size; i++) Fprintf(strm, "\t%f ", fea_vec[i]);
    Fprintf(strm, "\n");

}



void
vec_cy_f(dst, src, n)
float  *dst, *src;	/*pointers to destination and source vectors*/
long n;		/*size of vectors*/
/*
 *This routine copies src to dst.
 */
{
    while (n-- != 0)
	*dst++ = *src++;
}



